src/components/cylc/Graph.vue

```diff
@@ -0,0 +1,1284 @@
+  expandCueImage: undefined, // image of expand icon if undefined draw regular expand cue
+  collapseCueImage: undefined, // image of collapse icon if undefined draw regular collapse cue
+  expandCollapseCueSensitivity: 1 // sensitivity of expand-collapse cues
+}
+
+const expandCollapseOptionsCola = {
+  layoutBy: {
+    name: 'cose-bilkent',
+    animate: 'end',
+    randomize: false,
+    fit: false
+  },
+  // recommended usage: use cose-bilkent layout with randomize: false to preserve mental map upon expand/collapse
+  fisheye: false, // whether to perform fisheye view after expand/collapse you can specify a function too
+  animate: false, // whether to animate on drawing changes you can specify a function too
+  ready: function () {}, // callback when expand/collapse initialized
+  undoable: true, // and if undoRedoExtension exists,
+  cueEnabled: true, // Whether cues are enabled
+  expandCollapseCuePosition: 'top-left', // default cue position is top left you can specify a function per node too
+  expandCollapseCueSize: 20, // size of expand-collapse cue
+  expandCollapseCueLineSize: 16, // size of lines used for drawing plus-minus icons
+  expandCueImage: undefined, // image of expand icon if undefined draw regular expand cue
+  collapseCueImage: undefined, // image of collapse icon if undefined draw regular collapse cue
+  expandCollapseCueSensitivity: 1 // sensitivity of expand-collapse cues
+}
+
+export default {
+  name: 'Graph',
+  props: ['workflowid'],
+  data: function () {
+    return {
+      isConnected: false,
+      config,
+      elements,
+      graphData: {
+        nodes: [],
+        edges: []
+      },
+      i: 1,
+      // vue-spinner
+      color: '#5e9aff',
+      height: '35px',
+      width: '4px',
+      margin: '2px',
+      radius: '2px',
+      size: '1em',
+      loading: true,
+      // layout variables
+      status: 'pending',
+      layoutName: 'dagre',
+      layoutStopped: true,
+      layoutReady: false,
+      // elements
+      nodesEdges: [],
+      workflows: [],
+      edges: [],
+      subscriptions: {},
+      workflowId: ''
+    }
+  },
+  watch: {
+    graphData: debounce(function (newVal) {
+      this.debouncer.call()
+    }, 100),
+
+    layoutName: {
+      handler (value) {
+        this.updateLayout(value)
+      }
+    },
+
+    workflows: {
+      handler: function (newval, oldval) {
+        this.workflowUpdated(newval)
+      },
+      deep: true
+    }
+  },
+
+  mixins: [mixin],
+
+  metaInfo () {
+    return {
+      title: this.getPageTitle('App.graph', { name: this.workflowId })
+    }
+  },
+
+  beforeRouteLeave (to, from, next) {
+    if (tippy) {
+      tippy.hide()
+    }
+    next()
+  },
+
+  beforeDestroy () {
+    workflowService.unregister(this)
+  },
+
+  mounted () {
+    console.log(`MOUNTED called, status: ${this.status}`)
+    this.handleMounted()
+    this.$store.watch((store) => {
```

Comment:
I prefer using the watch paradigm, and this will also be used when we move to websockets

Reply:
The issue here is that the Tree component is adopting a different style. And I would be happy with either, but not both styles in the project, as I think it would be easier for maintainers if the components followed a certain pattern.

If this is better for the websockets, then I think we would have to discuss it when implementing websockets, and raise a PR to possibly fix it in both the Tree and the Graph components.

I really don't have a strong preference on either, but I use [this part of the Vue.js docs](https://vuejs.org/v2/guide/computed.html#Computed-vs-Watched-Property) as reference whenever I use a `watch` (which I just did recently in the Tree, then found out that it was not necessary and reverted again *):

> (...) it is tempting to overuse watch - especially if you are coming from an AngularJS background. However, it is often a better idea to use a computed property rather than an imperative watch callback.

So I think if we have the option to follow the documentation recommendations, we should. But as you pointed that it may be better for websockets, my suggestion here would be to be consistent with what we have for now in the other components, and move it to follow up pull requests.

* https://github.com/cylc/cylc-ui/pull/251/files#diff-d70e3541f48c5fdb3978c257a8726eceL38-R40 & https://github.com/cylc/cylc-ui/pull/251#issuecomment-532932593

Reply:
The problem is the data is very nested. The Vuex recommendation is to have a flatter data structure, and as I have to reconstruct new node and edges data it works better to watch the store using `deep`. I'll take a look at i though :-)

Reply:
In that case, it could be that we both need to use `watch` + `deep`. The tree component also has a structure with multiple levels (tree hierarchy) :+1:

Reply:
or have flatter data